src/iosrv/srv.c

```diff
@@ -402,6 +402,55 @@ wait_all_exited(struct dss_xstream *dx)
 	D_DEBUG(DB_TRACE, "XS(%d) drained ULTs.\n", dx->dx_xs_id);
 }
 
+/** data structure passed to progress thread when leveraging hyperthread */
+struct dss_progress_info {
+	struct dss_xstream	*dx;
+	struct dss_module_info	*dmi;
+	bool			 should_stop;
+};
+
+/** handler of the progress thread when using SMT */
+static void *
+dss_progress_handler(void *arg)
+{
+	struct dss_progress_info	*info = (struct dss_progress_info *)arg;
+	int				 rc;
+
+	/**
+	 * Set cpu affinity
+	 */
+	rc = hwloc_set_cpubind(dss_topo, info->dx->dx_cpuset_net,
+			       HWLOC_CPUBIND_THREAD);
+	if (rc)
+		D_ERROR("failed to set cpu net affinity: %d\n", errno);
+
+	/**
+	 * Set memory affinity, but fail silently if it does not work since some
+	 * systems return ENOSYS.
+	 */
+	rc = hwloc_set_membind(dss_topo, info->dx->dx_cpuset_net,
+			       HWLOC_MEMBIND_BIND, HWLOC_MEMBIND_THREAD);
+	if (rc)
+		D_DEBUG(DB_TRACE, "failed to set memory net affinity: %d\n",
+			errno);
+
+	while (!info->should_stop) {
+		/**
+		 * could eventually afford to pass a non-null timeout here
+		 * since the thread is dedicated to progress
+		 */
+		rc = crt_progress(info->dmi->dmi_ctx, 0);
```

Comment:
We have a barrier to prevent requests processing until all xstreams started (see dss_xstreams_open_barrier(), and the comment for barrier in dss_srv_handler()).

Reply:
Good catch.